### ISSUE TITLE
IDE0039 C# Use local function wrongly reported

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#28936]
IDE0039 (C# Use local function) should not be reported when a delegate is being created to be used as a delegate and not directly invoked.

What's the difference between this:
```C#
public void M()
{
    Action a = () => {};
    N(a);
}
```
and this?
```C#
public void M()
{
    void a() => {};
    N(a);
}
```
_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/301282/ide0039-c-use-local-function-wrongly-reported.html
VSTS ticketId: 655348_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by CyrusNajmabadi** [comment]
> IDE0039 (C# Use local function) should not be reported when a delegate is being created to be used as a delegate and not directly invoked.

Sorry, the issue is missing the "why" explanation here.  Why should it not be reported?

**Comment by JeffJ-DG** [comment]
Not OP, but what is the purpose of the IDE0039 (C# Use local function) recommendation in the first place?

**Comment by CyrusNajmabadi** [comment]
> Not OP, but what is the purpose of the IDE0039 (C# Use local function) recommendation in the first place?

In general, local functions can be better for performance since the compiler can optimize the case better where the function isn't captured into a delegate.

There are also benefits like being able to place them in more readable locations (i.e. not needing to be interspersed with the other statements of the method).  

Basically, for many cases, they're better than lambdas (which is why they were introduced in the first place).  So the feature attempts to let you know about this so you can move code easily over to using it.

**Comment by sharwell** [comment]
💭 Perhaps we should have one diagnostic for cases where conversion to a local function _eliminates_ the delegate, and another diagnostic for cases where it's just a different way to write the same thing.

**Comment by martingalvan** [comment]
> 
> 
> > Not OP, but what is the purpose of the IDE0039 (C# Use local function) recommendation in the first place?
> 
> In general, local functions can be better for performance since the compiler can optimize the case better where the function isn't captured into a delegate.
> 
> There are also benefits like being able to place them in more readable locations (i.e. not needing to be interspersed with the other statements of the method).
> 
> Basically, for many cases, they're better than lambdas (which is why they were introduced in the first place). So the feature attempts to let you know about this so you can move code easily over to using it.

Why do lambdas exist in the language, then?

**Comment by CyrusNajmabadi** [comment]
> Why do lambdas exist in the language, then?

Because while local-functions are better in many cases, lambdas are also better in other cases.  Examples being, but not limited to, things like terseness and ergonomics.  Lambdas don't need names.  Can have inference applied to them, etc.

**Comment by martingalvan** [comment]
Correction: why do multiple-statement lambdas exist in the language, then? Why not do as Python does, where lambdas are always single-line, and if you want something more complex you just define a function?

**Comment by CyrusNajmabadi** [comment]
> Correction: why do multiple-statement lambdas exist in the language, then?

Because we think multiple-statement lambdas are a totally reasonable thing for a person to use.  In roslyn we probably have hundreds to thousands of uses of them :)

> Why not do as Python does

**Comment by martingalvan** [comment]
> 
> 
> > Correction: why do multiple-statement lambdas exist in the language, then?
> 
> Because we think multiple-statement lambdas are a totally reasonable thing for a person to use. In roslyn we probably have hundreds to thousands of uses of them :)
> 
> > Why not do as Python does
> 
> We were not influenced by python here when designing our lambdas. Being able to perform many steps in a lambda was always seen as a desirable thing, and not something to restrict. So we designed them to support this from teh first version of hte language we released them in.
> 
> > and if you want something more complex you just define a function?
> 
> We don't view it as either/or. They are different tools for different circumstances. Local functions are not always the preferred way to do things, and neither are lambdas. Both exist because, as mentioned above, there are different use cases where either can be more appropriate.

Then why does the main, officially supported linter flag multiple-statement lambdas as something that is wrong?

**Comment by CyrusNajmabadi** [comment]
> Then why does the main, officially supported linter flag multiple-statement lambdas as something that is wrong?

It does not.  It simply suggests that you could use a local function here instead of a local variable assigned a lambda.  The former is generally more efficient than the latter here.  But this has no bearing on there still being a need for lambdas in the language.

In other words, the suggestions is about this *specific pattern*.  It is not a message to not use lambdas (esp. in all the places they are extremely valuable) and only use local functions.  It's simply that there is a pattern that is considered slightly more idiomatic (and which has better perf) *for this exactly this case*.

We do this in many places in the language.  There may be new features that subsume a *subset* of cases where older features may already be used.  When we encounter this, we suggest using that feature if appropriate.

**Comment by CyrusNajmabadi** [comment]
Closing out as by design.
